Sim/SimDelphesInterface/src/DelphesSimulation.cpp

```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+              << " Vx: "       << setw(10) << particle.EndVertex().Position().X
+              << " Vy: "       << setw(10) << particle.EndVertex().Position().Y
+              << " Vz: "       << setw(10) << particle.EndVertex().Position().Z
+              << " T: "        << setw(10) << particle.EndVertex().Ctau();
+    }
+    std::cout << std::endl;*/
+  }
+}   
 
-  Candidate * cand;
-  for(int j = 0; j < Input->GetEntries(); ++j) {
+void DelphesSimulation::ConvertTracks(const TObjArray*  Input,
+                                      fcc::ParticleCollection*   colParticles,
+                                      fcc::MCParticleCollection* colMCParticles,
+                                      fcc::ParticleMCParticleAssociationCollection* ascColParticlesToMC)
+{
+  Candidate* cand = nullptr;
+  for(int j=0; j<Input->GetEntries(); j++) {
 
     cand = static_cast<Candidate *>(Input->At(j));
-    fcc::GenJet outptc = coll->create();
-    fcc::BareJet& core = outptc.Core();
-    core.Area     = cand->Area.Mag();
-    core.P4.Px    = (double) cand->Momentum.X();
-    core.P4.Py    = (double) cand->Momentum.Y();
-    core.P4.Pz    = (double) cand->Momentum.Z();
-    core.P4.Mass  = (double) cand->Mass ;
-  }
+
+    auto particle = colParticles->create();
+    auto relation = ascColParticlesToMC->create();
+
+    auto barePart      = new fcc::BareParticle();
+    barePart->Type     = cand->PID;
+    barePart->Status   = cand->Status;
+    barePart->P4.Px    = cand->Momentum.Px();
+    barePart->P4.Py    = cand->Momentum.Py();
+    barePart->P4.Pz    = cand->Momentum.Pz();
+    barePart->P4.Mass  = cand->Momentum.M();
+    barePart->Charge   = cand->Charge;
+    barePart->Vertex.X = cand->Position.X();
+    barePart->Vertex.Y = cand->Position.Y();
+    barePart->Vertex.Z = cand->Position.Z();
+
+    // Reference to MC - Delphes holds references to all objects related to the <T> object, only one relates to MC particle
+    Candidate* refCand = nullptr;
+    int idRefMCPart    = -1;
+    if (cand->GetCandidates()->GetEntries()>0) {
+
+      refCand     = static_cast<Candidate*>(cand->GetCandidates()->At(0));
+      idRefMCPart = refCand->GetUniqueID()-1;     // Use C numbering from 0
+      if (idRefMCPart<colMCParticles->size()) {
+
+        barePart->Bits = ParticleStatus::Matched;
+        particle.Core(*barePart);
+        relation.Rec(particle);
+        relation.Sim(colMCParticles->at(idRefMCPart));
+      }
+      else {
+        barePart->Bits = ParticleStatus::Unmatched;
+        particle.Core(*barePart);
+        warning() << "Can't build relation from Electron/Muon/ChHadron to MC particle!" << std::endl;
+      }
+    }
+    else {
+      barePart->Bits = ParticleStatus::Unmatched;
+      particle.Core(*barePart);
+      warning() << "Can't build relation from Electron/Muon/ChHadron to MC particle!" << std::endl;
+    }
+    if (barePart!=nullptr) delete barePart;
+
+    // Test print
```

Comment:
Please remove these lines if they don't serve any purpose any more. Or make it a debug output.